File: v1/server/services/templates.ts
import Handlebars from "handlebars"
import { cpSync, existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs"
import { dirname, join, normalize, resolve, sep } from "node:path"
import { appsRepo } from "../db/apps"
import { configSetsRepo } from "../db/config-sets"
import { envVarsRepo } from "../db/env-vars"
import { templatesRepo } from "../db/templates"

function backupRoot(appId: number, sessionId: string) {
  return join(process.cwd(), "data", "backups", String(appId), sessionId)
}

function resolveSafePath(projectPath: string, relativePath: string): string {
  const cleaned = relativePath.replace(/\\/g, "/").replace(/^\/+/, "")
  const full = resolve(projectPath, cleaned)
  const root = resolve(projectPath)
  const normalizedFull = normalize(full)
  const normalizedRoot = normalize(root)
  if (
    normalizedFull !== normalizedRoot &&
    !normalizedFull.startsWith(normalizedRoot + sep)
  ) {
    throw new Error(`Template path escapes project directory: ${relativePath}`)
  }
  return full
}

export type TemplateApplyResult = {
  applied: string[]
  backupDir: string
}

export function applyTemplates(
  appId: number,
  sessionId: string
): TemplateApplyResult {
  const app = appsRepo.get(appId)
  if (!app) throw new Error("App not found")

  const set = configSetsRepo.resolveActive(appId)
  const templates = templatesRepo.listByConfigSet(set.id)
  const env = envVarsRepo.toRecord(set.id)
  const backupDir = backupRoot(appId, sessionId)
  const applied: string[] = []

  mkdirSync(backupDir, { recursive: true })

  try {
    for (const template of templates) {
      const targetPath = resolveSafePath(app.project_path, template.file_path)
      if (!existsSync(targetPath)) {
        throw new Error(`Target file does not exist: ${template.file_path}`)
      }

      const backupPath = join(backupDir, template.file_path.replace(/\\/g, "/"))
      mkdirSync(dirname(backupPath), { recursive: true })
      cpSync(targetPath, backupPath)

      const compiled = Handlebars.compile(template.content, { noEscape: true })
      const rendered = compiled(env)
      writeFileSync(targetPath, rendered, "utf8")
      applied.push(template.file_path)
    }
  } catch (err) {
    restoreTemplates(appId, sessionId)
    throw err
  }

  return { applied, backupDir }
}

export function restoreTemplates(appId: number, sessionId: string): string[] {
  const app = appsRepo.get(appId)
  if (!app) return []

  const backupDir = backupRoot(appId, sessionId)
  if (!existsSync(backupDir)) return []

  const restored: string[] = []
  const set = configSetsRepo.resolveActive(appId)
  const templates = templatesRepo.listByConfigSet(set.id)

  for (const template of templates) {
    const backupPath = join(backupDir, template.file_path.replace(/\\/g, "/"))
    if (!existsSync(backupPath)) continue
    const targetPath = resolveSafePath(app.project_path, template.file_path)
    mkdirSync(dirname(targetPath), { recursive: true })
    const content = readFileSync(backupPath)
    writeFileSync(targetPath, content)
    restored.push(template.file_path)
  }

  rmSync(backupDir, { recursive: true, force: true })
  return restored
}

File: wails_backend/internal/services/templates.go
package services

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/aymerick/raymond"

	"wails_backend/internal/db"
	"wails_backend/internal/lib"
)

func renderHandlebars(content string, env map[string]string) (string, error) {
	tpl, err := raymond.Parse(content)
	if err != nil {
		return "", err
	}
	ctx := make(map[string]any, len(env))
	for k, v := range env {
		ctx[k] = raymond.SafeString(v)
	}
	return tpl.Exec(ctx)
}

func backupRoot(appID int64, sessionID string) string {
	return filepath.Join(db.DataDir(), "backups", strconv.FormatInt(appID, 10), sessionID)
}

func ApplyTemplates(ctx context.Context, d *db.DB, appID int64, sessionID string) error {
	app, err := d.GetAppT(ctx, appID)
	if err != nil {
		return err
	}
	set, err := d.ResolveActive(ctx, appID)
	if err != nil {
		return err
	}
	templates, err := d.ListTemplatesT(ctx, set.ID)
	if err != nil {
		return err
	}
	env, err := d.EnvToRecord(ctx, set.ID)
	if err != nil {
		return err
	}
	backupDir := backupRoot(appID, sessionID)
	if err := os.MkdirAll(backupDir, 0o755); err != nil {
		return err
	}
	for _, template := range templates {
		targetPath, err := lib.ResolveSafePath(app.ProjectPath, template.FilePath)
		if err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return err
		}
		if _, err := os.Stat(targetPath); err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return fmt.Errorf("Target file does not exist: %s", template.FilePath)
		}
		backupPath := filepath.Join(backupDir, filepath.FromSlash(strings.ReplaceAll(template.FilePath, "\\", "/")))
		if err := os.MkdirAll(filepath.Dir(backupPath), 0o755); err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return err
		}
		data, err := os.ReadFile(targetPath)
		if err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return err
		}
		if err := os.WriteFile(backupPath, data, 0o644); err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return err
		}
		rendered, err := renderHandlebars(template.Content, env)
		if err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return fmt.Errorf("failed to render template %s: %w", template.FilePath, err)
		}
		if err := os.WriteFile(targetPath, []byte(rendered), 0o644); err != nil {
			_ = RestoreTemplates(ctx, d, appID, sessionID)
			return err
		}
	}
	return nil
}

func RestoreTemplates(ctx context.Context, d *db.DB, appID int64, sessionID string) error {
	app, err := d.GetAppT(ctx, appID)
	if err != nil {
		return nil
	}
	backupDir := backupRoot(appID, sessionID)
	if _, err := os.Stat(backupDir); err != nil {
		return nil
	}
	set, err := d.ResolveActive(ctx, appID)
	if err != nil {
		return err
	}
	templates, err := d.ListTemplatesT(ctx, set.ID)
	if err != nil {
		return err
	}
	for _, template := range templates {
		backupPath := filepath.Join(backupDir, filepath.FromSlash(strings.ReplaceAll(template.FilePath, "\\", "/")))
		if _, err := os.Stat(backupPath); err != nil {
			continue
		}
		targetPath, err := lib.ResolveSafePath(app.ProjectPath, template.FilePath)
		if err != nil {
			continue
		}
		_ = os.MkdirAll(filepath.Dir(targetPath), 0o755)
		data, err := os.ReadFile(backupPath)
		if err != nil {
			continue
		}
		_ = os.WriteFile(targetPath, data, 0o644)
	}
	return os.RemoveAll(backupDir)
}
